pkg/common/log/log.go

```diff
@@ -0,0 +1,110 @@
+package log
+
+import (
+	"strings"
+
+	"github.com/go-logr/logr"
+	"go.uber.org/zap/zapcore"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+var (
+	Level LogLevel
+
+	baseLogger logr.Logger = ctrl.Log
+)
+
+type LogLevel zapcore.Level
+
+func (l *LogLevel) Debug() bool {
+	return *l == LogLevel(zapcore.DebugLevel)
+}
+
+func (l *LogLevel) String() string {
+	return zapcore.Level(*l).String()
+}
+
+// ToLogLevel converts a string to a log level.
+func ToLogLevel(level string) LogLevel {
+	var l zapcore.Level
+	_ = l.UnmarshalText([]byte(level))
+	return LogLevel(l)
+}
+
+// LogMode defines the log output mode.
+type LogMode int8
+
+const (
+	// LogModeProd is the log mode for production.
+	LogModeProd LogMode = iota
+	// LogModeDev is for more human-readable outputs, extra stack traces
+	// and logging info. (aka Zap's "development config".)
+	LogModeDev
+)
+
+func (f *LogMode) String() string {
+	switch *f {
+	case LogModeProd:
+		return "production"
+	case LogModeDev:
+		return "development"
+	default:
+		return "unknown"
+	}
+}
+
+// ToLogMode converts a string to a log mode.
+// Use either 'production' for `LogModeProd` or 'development' for `LogModeDev`.
+func ToLogMode(mode string) LogMode {
+	switch strings.ToLower(mode) {
+	case "production":
+		return LogModeProd
+	case "development":
+		return LogModeDev
+	default:
+		panic("unknown log mode")
+	}
+}
+
+// Setup sets up a logger with the given level and output mode.
+func Setup(level LogLevel, mode LogMode) logr.Logger {
+	logger := NewLogger(level, mode)
+
+	ctrl.SetLogger(logger)
+	klog.SetLogger(logger)
+	Level = level
+
+	Logger().Info("setting instance base logger", "min level", level.String(), "mode", mode.String())
+
+	return logger
+}
+
+// NewLogger returns a new logger with the given level and output mode.
+func NewLogger(level LogLevel, mode LogMode) logr.Logger {
+	return zap.New(
+		zap.Level(zapcore.Level(level)),
+		zap.UseDevMode(mode == LogModeDev),
+	)
+}
+
+// Logger returns the logger with the current log level.
+func Logger() logr.Logger {
```

Comment:
It is not obvious but the logger is actually _configured_. When we `log.Setup(logger)`, the configured logger becomes the one ultimately fulfilled by `ctrl.Log` and used by `k8s.io/klog`; then `baseLogger` points to that logger and `Logger()` basically returns `baseLogger`. I'll try to make this clearer.

Alternatively, I thought about injecting a preconfigured logger along into the stack of structs involved in every external auth request (auth service → auth pipeline → auth config → evaluators call), but then it became easier to just use the base logger; after all, apart from the ext authz call, we also have the reconcilers → new configs/evaluators flow.

Injecting the logger would simplify the "useful things" I've mentioned in https://github.com/Kuadrant/authorino/pull/172#discussion_r731716281.

Reply:
I've tried to enhance this a little bit, based on @eguzki suggestions:
- [x] Singleton `"github.com/kuadrant/authorino/pkg/common/log".Log` configured by invoking `SetLogger`, similarly to the sigs.k8s.io/controller-runtime's delegation logger.
- [x] Authorino preconfigured logger as a proper delegation logger (instead of relying on the singleton always pointing to `ctrl.Log`), and storing the log options (level and mode)
- [x] Loggers injected along the call stack: hierarchically chaining log names, passing along preconfigured values (e.g. "request id")